flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/sink/SocketClientSink.java

```diff
@@ -73,13 +90,56 @@ public void intializeConnection() {
 	 *			The incoming data
 	 */
 	@Override
-	public void invoke(IN value) {
+	public void invoke(IN value) throws Exception {
 		byte[] msg = schema.serialize(value);
 		try {
 			dataOutputStream.write(msg);
 		} catch (IOException e) {
-			throw new RuntimeException("Cannot send message " + value.toString() +
-					" to socket server at " + hostName + ":" + port, e);
+			LOG.error("Cannot send message " + value.toString() +
+					" to socket server at " + hostName + ":" + port + ". Caused by " + e.toString() +
+					". Trying to reconnect.");
+			retries = 0;
+			boolean success = false;
+			while ((retries < maxRetry || retryForever) && !success && isRunning){
+				try {
+
+					if (dataOutputStream != null) {
+						dataOutputStream.close();
+					}
+
+					if (client != null && !client.isClosed()) {
+						client.close();
+					}
+
+					retries++;
+
+					client = new Socket(hostName, port);
+					dataOutputStream = new DataOutputStream(client.getOutputStream());
+					dataOutputStream.write(msg);
+					success = true;
+
+				} catch(IOException ee) {
+					LOG.error("Reconnect to socket server and send message failed. Caused by " +
+							ee.toString() + ". Retry time(s):" + retries);
+
+					if (lock == null) {
+						lock = new Object();
+					}
+
+					try {
+						synchronized (lock) {
+							lock.wait(CONNECTION_RETRY_SLEEP);
+						}
+					} catch(InterruptedException eee) {
+						LOG.error(eee.toString());
```

Comment:
I think there is no need to log this. Interrupting the thread usually means that it has been shut down.

If you want to log exceptions, log them via ("message", exeption), here `LOG.error("Reconnect delay interrupted", e);`